### PR TITLE
Eterm seggregation

### PR DESCRIPTION
**Pull request by Jacarte**
As we discussed previously, we have now the `enodes` in the peephole mutator separated by type. The current implementation is for `i32` and `i64` instructions. 

The code needs a little bit more of depuration, so I will keep this PR as a draft, meanwhile cc @fitzgen :)

### CODE REVIEW

**Comment by fitzgen** [comment]
> As we discussed previously, we have now the `enodes` in the peephole mutator separated by type. The current implementation is for `i32` and `i64` instructions.

The plan is still to remove all the polymorphism from e-classes though, right?

**Comment by Jacarte** [comment]
> > As we discussed previously, we have now the `enodes` in the peephole mutator separated by type. The current implementation is for `i32` and `i64` instructions.
> 
> The plan is still to remove all the polymorphism from e-classes though, right?

Yep

**Comment by Jacarte** [comment]
Polymorphism is removed from all enodes. This means more hand work for rewriting rules writing, for example, now instead of having one rule for commutativeness `(add ?x ?y) <=> (add ?y ?x)` we will need to specify the typed operators like 

```
(i32add ?x ?y) <=> (i32add ?y ?x)
(i64add ?y ?x) <=> (i64add ?y ?x)
```
In practice the size of the `egraph` is maintained. The complexity of the API is decreased as well. 

On the other hand, for artificially created nodes, for example, neutral rewriting rules like `?x => (add ?x 0)`, some conditionals are needed. Since the add operator has a type the variable `x` should be checked to have the same type. This can be done by using conditionals during the creation of the rewriting rules. For example

```rust
?x => (i32add ?x 0) if is_type("?x", PrimitiveInfoType::I32)
``` 

Notice that the previous rewriting rule needs to be written for the four types in case we wanted them for `i64, f32 and f64` as well

**Comment by Jacarte** [comment]
Fixes #388

**Comment by Jacarte** [comment]
> Polymorphism is removed from all enodes. This means more hand work for rewriting rules writing, for example, now instead of having one rule for commutativeness `(add ?x ?y) <=> (add ?y ?x)` we will need to specify the typed operators like
> 
> ```
> (i32add ?x ?y) <=> (i32add ?y ?x)
> (i64add ?y ?x) <=> (i64add ?y ?x)
> ```
> 
> In practice the size of the `egraph` is maintained. The complexity of the API is decreased as well.
> 
> On the other hand, for artificially created nodes, for example, neutral rewriting rules like `?x => (add ?x 0)`, some conditionals are needed. Since the add operator has a type the variable `x` should be checked to have the same type. This can be done by using conditionals during the creation of the rewriting rules. For example
> 
> ```rust
> ?x => (i32add ?x 0) if is_type("?x", PrimitiveInfoType::I32)
> ```
> 
> Notice that the previous rewriting rule needs to be written for the four types in case we wanted them for `i64, f32 and f64` as well

On interesting issue here, as soon as we start to grow the `language` definition with `egg`, it reaches `recursion limit reached while expanding $crate::__define_language! ` for the `define_language` macro.

**Comment by fitzgen** [comment]
> Is it recommended to use `#![recursion_limit = "X"]` here?

Yeah we should be able to use that to bump the recursion limit.

File: crates/wasm-mutate/src/mutators/peephole/eggsy/encoder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+                TraversalEvent::Exit => {
+                    match &entry.operator {
+                        Lang::I32Add(_) => {
+                            newfunc.instruction(&Instruction::I32Add);
+                        }
+                        Lang::I64Add(_) => {
+                            newfunc.instruction(&Instruction::I64Add);
+                        }
+                        Lang::I32Sub(_) => {
+                            newfunc.instruction(&Instruction::I32Sub);
```
Comment (fitzgen) [inline crates/wasm-mutate/src/mutators/peephole/eggsy/encoder/mod.rs:59]: Wait -- how is this different from `expr2wasm`? Why do we have two versions of `Lang` to Wasm encoding?

**Review by fitzgen — COMMENTED** [review]
Nice! A lot simpler now. Just a few comments below.

File: crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs [hunk 2]
```diff
@@ -407,27 +409,56 @@ impl<'a> DFGBuilder {
...
                         color,
                         locals[*local_index as usize].clone(),
                     );
+
+                    self.parents[arg] = idx as i32;
                 }
                 Operator::GlobalGet { global_index } => {
                     self.push_node(
-                        StackType::GlobalGet(*global_index),
+                        Lang::Arg(*global_index as u64),
```
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs:431]: Actually, we do so. We only push and pop the arg right away to use it as the `global_idx` operand for the `global.get`

File: crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs [hunk 3]
```diff
@@ -407,27 +409,56 @@ impl<'a> DFGBuilder {
...
+                        idx,
+                        vec![],
+                        color,
+                        PrimitiveTypeInfo::I32,
+                    );
+                    let arg = self.pop_operand(idx, false);
+
+                    let idx = self.push_node(
+                        Lang::GlobalGet([Id::from(arg)]),
```
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs:440]: Honestly, I think the best way we have to flatten this is by using the `Arg` node. Regarding possible debugging, numeric nodes will be rendered as the value, so, we will have `Local::GlobalGet i` after all.

For semantically equivalent transformations, I agree that the index should not change. However, for a real "chaos" mode the index operand can be replaced by any value.

File: crates/wasm-mutate/src/mutators/peephole/eggsy/encoder/mod.rs (hunk 1 above)
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/eggsy/encoder/mod.rs:59]: Working on it, after removing our "intermediate" `StackEntry` conversion a lot of code start to be repetitive, and this case is probably something that we can remove.

File: crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs:438]: The other option is to make our own "tokens". The egg API allow us to define custom tokens if they implement `FromStr` trait

[the remark this answers is not in the document]
Comment (fitzgen) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs:438]: Can you share a link to the "tokens" stuff?

[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs:438]: Sure :), https://github.com/egraphs-good/egg/blob/afbd4821f0cf1725c6e72a63612635f14489c0cf/src/macros.rs#L48

File: crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs [hunk 4]
```diff
@@ -485,12 +536,101 @@ impl<'a> DFGBuilder {
...
+                    let align = self.pop_operand(idx, false);
+
+                    self.push_node(
+                        Lang::Arg(memarg.memory as u64),
+                        idx,
+                        vec![],
+                        color,
+                        PrimitiveTypeInfo::I32,
+                    );
+                    let memory = self.pop_operand(idx, false);
```
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs]: Inmediates issue solved with manual `Language` implementation for `Lang`

File: crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/peephole/dfg/mod.rs]: Manual implementation for `egg::Language trait`

**Comment by Jacarte** [comment]
cc @fitzgen

**Review by fitzgen — APPROVED** [review]
Looks great! Thanks for implementing `Lang` by hand, I think we definitely outgrew the macro.

**Pull request merged**